share/spack/win-compiler-wrapper/cl.cxx

```diff
@@ -0,0 +1,406 @@
+    STARTUPINFOW siStartInfo;
+    ZeroMemory( &piProcInfo, sizeof(PROCESS_INFORMATION) );
+
+    // Set up members of the STARTUPINFO structure.
+    // This structure specifies the STDIN and STDOUT handles for redirection.
+    ZeroMemory( &siStartInfo, sizeof(STARTUPINFOW) );
+    siStartInfo.cb = sizeof(STARTUPINFOW);
+    siStartInfo.hStdError = this->ChildStdOut_Wd;
+    siStartInfo.hStdOutput = this->ChildStdOut_Wd;
+    siStartInfo.dwFlags |= STARTF_USESTDHANDLES;
+    this->procInfo = piProcInfo;
+    this->startInfo= siStartInfo;
+}
+
+bool ToolChainInvocation::pipeChildtoStdOut(){
+    DWORD dwRead, dwWritten;
+    CHAR chBuf[BUFSIZE];
+    BOOL bSuccess = TRUE;
+    HANDLE hParentStdOut = GetStdHandle(STD_OUTPUT_HANDLE);
+
+    for (;;)
+    {
+        bSuccess = ReadFile( this->ChildStdOut_Rd, chBuf, BUFSIZE, &dwRead, NULL);
+        if( ! bSuccess || dwRead == 0 ) break;
+
+        bSuccess = WriteFile(hParentStdOut, chBuf,
+                            dwRead, &dwWritten, NULL);
+        if (! bSuccess ) break;
+    }
+    return bSuccess;
+}
+
+void ToolChainInvocation::executeToolChainChild(){
+    LPVOID lpMsgBuf;
+    wchar_t * commandLine = &ConvertAnsiToWide(this->composeCLI())[0];
+    if(! CreateProcessW(
+        ConvertAnsiToWide(this->spackCommand).c_str(),
+        commandLine,
+        NULL,
+        NULL,
+        TRUE,
+        0,
+        NULL,
+        NULL,
+        &this->startInfo,
+        &this->procInfo)
+    )
+        // Handle errors coming from creating of child proc
+        FormatMessage(
+            FORMAT_MESSAGE_ALLOCATE_BUFFER |
+            FORMAT_MESSAGE_FROM_SYSTEM |
+            FORMAT_MESSAGE_IGNORE_INSERTS,
+            NULL,
+            GetLastError(),
+            MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+            (LPTSTR) &lpMsgBuf,
+            0, NULL
+        );
+        throw SpackException((char *)lpMsgBuf);
+    // We've suceeded in kicking off the toolchain run
+    // Explicitly close write handle to child proc stdout
+    // as it is no longer needed and if we do not when cannot
+    // determine when child proc is done
+    CloseHandle(this->ChildStdOut_Wd);
+}
+
+void ToolChainInvocation::createChildPipes(){
+    SECURITY_ATTRIBUTES saAttr;
+    // Set the bInheritHandle flag so pipe handles are inherited.
+    saAttr.nLength = sizeof(SECURITY_ATTRIBUTES);
+    saAttr.bInheritHandle = TRUE;
+    saAttr.lpSecurityDescriptor = NULL;
+
+    if(!CreatePipe(&this->ChildStdOut_Rd, &this->ChildStdOut_Wd, &saAttr, 0))
+        throw SpackException("Could not create Child Pipe");
+    if ( !SetHandleInformation(ChildStdOut_Rd, HANDLE_FLAG_INHERIT, 0) )
+        throw SpackException("Child pipe handle inappropriately inhereited");
+}
+
+std::string ToolChainInvocation::composeIncludeArg(std::string &include){
+    return "/I" + include;
+}
+
+std::string ToolChainInvocation::composeLibPathArg(std::string &libPath){
+    return "/LIBPATH:" + libPath;
+}
```

Comment:
This appears to read in e.g. `SPACK_LINK_DIRS`, but does not add them to the invocation the way the POSIX wrapper does (i.e. turning them into `-L`. Likewise for `SPACK_INCLUDE_DIRS`

Reply:
See `interpolateSpackEnv` where I call `addExtraLibPaths` on the value read from the env in `loadSpackEnvState` where I then add them to `this->libDirArgs` and then pass that to the command line via `composeCLI`